### PR TITLE
Device handler for GE Color Effects remote control

### PR DESCRIPTION
**Pull request by lukecyca**
See also #569 for captures for this device

### CODE REVIEW

File: src/devices/ge_coloreffects.c [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+ */
+
+#include "rtl_433.h"
+#include "util.h"
+
+
+// Frame preamble:
+// 11001100 11001100 11001100 11001100 11001100 11111111 00000000
+// c   c    c   c    c   c    c   c    c   c    f   f    0   0
+static const unsigned char preamble_pattern[7] = { 0xcc, 0xcc, 0xcc, 0xcc, 0xcc, 0xff, 0x00};
```
Comment (zuckschwerdt) [inline src/devices/ge_coloreffects.c]: use a shorter preamble to make the reception more robust, i.e. `{0xcc, 0xff, 0x00}`

File: src/devices/ge_coloreffects.c [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+}
+
+static int ge_coloreffects_callback(bitbuffer_t *bitbuffer) {
+    unsigned bitpos = 0;
+    int events = 0;
+
+    // Find a preamble with enough bits after it that it could be a complete packet
+    // (if the device id and command were all zeros)
+    while ((bitpos = bitbuffer_search(bitbuffer, 0, bitpos, (uint8_t *)&preamble_pattern, 56)) + 34 <=
```
Comment (zuckschwerdt) [inline src/devices/ge_coloreffects.c]: you'd expect 78 bits here (24 + 27*2) with a 24 bit match (s.a.), i.e. `+ 78`

File: src/devices/ge_coloreffects.c [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+     *   One zero bit
+     */
+    
+    // Frame should be 17 decoded bits (not including preamble)
+    if (packet_bits.bits_per_row[0] != 17) {
+      return 0;
+    }
+    
+    // First two bits must be 0
+    if (bit(packet_bits.bb[0], 0) != 0 || bit(packet_bits.bb[0], 1) != 0) {
```
Comment (zuckschwerdt) [inline src/devices/ge_coloreffects.c]: `packet_bits.bb[0] & 0xc0 == 0` would be faster

File: src/devices/ge_coloreffects.c [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    }
+    
+    // Last bit must be 0
+    if (bit(packet_bits.bb[0], 16) != 0) {
+      return 0;
+    }
+    
+    // Extract device ID
+    // First two bits are zero, so we'll just take the entire first byte
+    bitbuffer_extract_bytes(&packet_bits, 0, 0, &device_id, 8);
```
Comment (zuckschwerdt) [inline src/devices/ge_coloreffects.c]: plain `packet_bits.bb[0]` would do

File: src/devices/ge_coloreffects.c [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+
+    // Extract command from the second byte
+    bitbuffer_extract_bytes(&packet_bits, 0, 8, &command, 8);
+    
+    // Format data
+    data = data_make(
```
Comment (zuckschwerdt) [inline src/devices/ge_coloreffects.c]: add `time` and `model` too

File: src/devices/ge_coloreffects.c [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+        "device",            "",     DATA_STRING, "GE Color Effects Remote",
```
Comment (zuckschwerdt) [inline src/devices/ge_coloreffects.c]: it's custom to name this `model`

File: src/devices/ge_coloreffects.c [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+        "device_id",         "",     DATA_FORMAT, "0x%x", DATA_INT, device_id,
```
Comment (zuckschwerdt) [inline src/devices/ge_coloreffects.c]: it's custom to name this just `id`

**Comment by zuckschwerdt** [comment]
Looks good. Also add a PR for the samples files to https://github.com/merbanan/rtl_433_tests

**Comment by merbanan** [comment]
@dmcwhinnie We are missing samples, if you submit some we can merge the code.

**Comment by lukecyca** [comment]
Sorry that I left this incomplete. Unfortunately I have no time to move this further right now.

There are sample captures attached to #569. I believe all that needs to happen is make a PR to integrate those into the rtl_433_tests repo. And resolve the conflicts in this PR.

Let me know if I can help and I will try to help get this finished.

**Pull request merged**